[CI] StatelessIndexCommitListenerIT testRefIsNotReleasedUponNotificationFailure failing

**Build Scans:**
- [elasticsearch-intake #39446 / part2](https://gradle-enterprise.elastic.co/s/hxnfjwtotzcki)
- [elasticsearch-intake #39444 / part2](https://gradle-enterprise.elastic.co/s/5qijbljlakr4s)
- [elasticsearch-pull-request #144549 / part-2](https://gradle-enterprise.elastic.co/s/hirfagwq377ng)
- [elasticsearch-pull-request #144549 / part-2](https://gradle-enterprise.elastic.co/s/rq5l7xsjhu5ry)
- [elasticsearch-pull-request #144542 / part-2](https://gradle-enterprise.elastic.co/s/ihwgv5zrpoc5q)
- [elasticsearch-pull-request #144538 / part-2](https://gradle-enterprise.elastic.co/s/u6lagxwpik6ca)
- [elasticsearch-intake #39195 / part2](https://gradle-enterprise.elastic.co/s/riyx3mbrk3dxg)

**Reproduction Line:**
```
./gradlew ":x-pack:plugin:stateless:internalClusterTest" --tests "org.elasticsearch.xpack.stateless.StatelessIndexCommitListenerIT.testRefIsNotReleasedUponNotificationFailure" -Dtests.seed=E42583D22A1974A1 -Dtests.locale=es-IC -Dtests.timezone=EST5EDT -Druntime.java=26
```

**Applicable branches:**
main

**Reproduces locally?:**
N/A

**Failure History:**
[See dashboard](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupState:(initialChildControlState:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:task.keyword,order:0,selectedOptions:!(),title:'GradleTask',type:optionsListControl),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:className.keyword,order:1,selectedOptions:!(org.elasticsearch.xpack.stateless.StatelessIndexCommitListenerIT),title:'Suite',type:optionsListControl),'144933da-5c1b-4257-a969-7f43455a7901':(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,fieldName:name.keyword,order:2,selectedOptions:!('testRefIsNotReleasedUponNotificationFailure'),title:'Test',type:optionsListControl)))))

**Failure Message:**
```
java.lang.AssertionError: IndexShard instance not found for shard [xilxummdlk][0] on nodes with [index] role
```

**Issue Reasons:**
- [main] 1 consecutive failures in test testRefIsNotReleasedUponNotificationFailure
- [main] 2 consecutive failures in step part2
- [main] 2 consecutive failures in pipeline elasticsearch-intake
- [main] 7 failures in test testRefIsNotReleasedUponNotificationFailure (8.1% fail rate in 86 executions)
- [main] 3 failures in step part2 (23.1% fail rate in 13 executions)
- [main] 4 failures in step part-2 (5.5% fail rate in 73 executions)
- [main] 3 failures in pipeline elasticsearch-intake (23.1% fail rate in 13 executions)
- [main] 3 failures in pipeline elasticsearch-pull-request (4.3% fail rate in 70 executions)

**Note:**
This issue was created using new test triage automation. Please report issues or feedback to es-delivery.

--- a/muted-tests.yml
+++ b/muted-tests.yml
@@ -497,9 +497,6 @@ tests:
 - class: org.elasticsearch.xpack.dlm.frozen.DLMFrozenTransitionIT
   method: testEndToEndFrozenTransition
   issue: https://github.com/elastic/elasticsearch/issues/148467
-- class: org.elasticsearch.xpack.stateless.StatelessIndexCommitListenerIT
-  method: testRefIsNotReleasedUponNotificationFailure
-  issue: https://github.com/elastic/elasticsearch/issues/148471
 - class: org.elasticsearch.xpack.stateless.cache.SearchCommitPrefetcherIT
   method: testCommitPrefetching
   issue: https://github.com/elastic/elasticsearch/issues/148476

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessIndexCommitListenerIT.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessIndexCommitListenerIT.java
@@ -24,6 +24,8 @@ import org.elasticsearch.plugins.PluginsService;
 import org.elasticsearch.xpack.stateless.commits.CommitBCCResolver;
 import org.elasticsearch.xpack.stateless.commits.IndexEngineLocalReaderListener;
 import org.elasticsearch.xpack.stateless.commits.StatelessCommitService;
+import org.elasticsearch.xpack.stateless.engine.IndexEngine;
+import org.elasticsearch.xpack.stateless.engine.IndexEngineTestUtils;
 import org.elasticsearch.xpack.stateless.engine.PrimaryTermAndGeneration;
 import org.elasticsearch.xpack.stateless.recovery.RegisterCommitResponse;
 import org.junit.After;
@@ -256,7 +258,9 @@ public class StatelessIndexCommitListenerIT extends AbstractStatelessPluginInteg
 
     @After
     public void teardownTest() {
-        assertAcked(client().admin().indices().prepareDelete(indexName));
+        if (indexName != null) {
+            assertAcked(client().admin().indices().prepareDelete(indexName));
+        }
         indexNode = null;
         searchNode = null;
         indexName = null;
@@ -420,8 +424,15 @@ public class StatelessIndexCommitListenerIT extends AbstractStatelessPluginInteg
             // Wait until the shard is marked as failed and then reallocated so we can
             // release the new commits once it becomes green again.
             ensureGreen(indexName);
+
+            // Delete the index to close shards and prevent new commits from being created
+            // before releasing retained commits, avoiding a race condition.
+            final var engine = asInstanceOf(IndexEngine.class, findIndexShard(indexName).withEngine(e -> e));
+            assertAcked(indicesAdmin().prepareDelete(indexName));
+            IndexEngineTestUtils.awaitClose(engine);
         } finally {
             plugin.listRetainedCommits(shardId).forEach(this::releaseCommit);
+            indexName = null;
         }
     }
 

--- a/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessIndexCommitListenerIT.java
+++ b/x-pack/plugin/stateless/src/internalClusterTest/java/org/elasticsearch/xpack/stateless/StatelessIndexCommitListenerIT.java
@@ -431,8 +431,11 @@ public class StatelessIndexCommitListenerIT extends AbstractStatelessPluginInteg
             assertAcked(indicesAdmin().prepareDelete(indexName));
             IndexEngineTestUtils.awaitClose(engine);
         } finally {
-            plugin.listRetainedCommits(shardId).forEach(this::releaseCommit);
-            indexName = null;
+            try {
+                plugin.listRetainedCommits(shardId).forEach(this::releaseCommit);
+            } finally {
+                indexName = null;
+            }
         }
     }
 

